### PR TITLE
wallet: Strengthen CWalletTx::RevalidateTransactions

### PR DESCRIPTION
**Pull request by jamescowens**
This extends RevalidateTransactions to check the transaction inputs and also check the contracts at both the vanilla and GRC level. Any invalid transactions are erased from the wallet.

I have noticed that stale transactions, or transactions with invalid contracts in the wallet database that have not been sent out previously leak out and are causing unnecessary misbehavior points. In a few cases nodes have been completely banned because of this.

This detects the invalid transactions and deletes them from both the wallet database and the mempool (if present).

### CODE REVIEW

**Comment by jamescowens** [comment]
@iFoggz I changed to you on this because @div72 has been busy on exams this past two weeks.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1277,8 +1267,35 @@ void CWallet::ResendWalletTransactions(bool fForce)
...
+        if (fInvalid) {
+            return error("%s: FetchInputs found invalid tx %s", __func__, tx.GetHash().ToString());
+        }
+        return error("%s: FetchInputs unable to fetch all inputs for tx %s", __func__, tx.GetHash().ToString());
+    }
+
+    // Validate any contracts published in the transaction:
+    if (!tx.GetContracts().empty()) {
+        if (!CheckContracts(tx, mapInputs, pindexBest->nHeight)) {
+            return error("%s: CheckContracts found invalid contract in tx %s", __func__, tx.GetHash().ToString());;
```
Comment (iFoggz) [inline src/wallet/wallet.cpp]: minor NIT an extra semicolon present.

**Review by iFoggz — COMMENTED** [review]
concept ACK.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (jamescowens) [reply src/wallet/wallet.cpp]: Fixed.

**Pull request merged**